File: frontend/src/api/posts.ts
import { api, toQuery } from './client'
import type { PostDetail, PostPage } from '../types'

export interface PostListParams {
  tag?: string // 쉼표 구분
  user?: string
  q?: string // 공백 구분
  page?: number
  limit?: number
}

export const fetchPosts = (params: PostListParams = {}) =>
  api<PostPage>(`/api/posts${toQuery(params)}`)

export const fetchMyPosts = (params: PostListParams = {}) =>
  api<PostPage>(`/api/users/me/posts${toQuery(params)}`)

export const fetchPost = (id: number) => api<PostDetail>(`/api/posts/${id}`)

export interface CreatePostBody {
  title: string
  content: string
  excerpt?: string
  tags?: string[]
}

export const createPost = (body: CreatePostBody) =>
  api<PostDetail>('/api/posts', { method: 'POST', body: JSON.stringify(body) })

export const updatePost = (id: number, body: Partial<CreatePostBody>) =>
  api<PostDetail>(`/api/posts/${id}`, { method: 'PATCH', body: JSON.stringify(body) })

export const deletePost = (id: number) => api<void>(`/api/posts/${id}`, { method: 'DELETE' })

export interface StarResponse {
  starred: boolean
  starCount: number
}

export const starPost = (id: number) => api<StarResponse>(`/api/posts/${id}/star`, { method: 'PUT' })

export const unstarPost = (id: number) =>
  api<StarResponse>(`/api/posts/${id}/star`, { method: 'DELETE' })

File: api/src/handlers/posts.rs
use axum::{
    Json,
    extract::{Path, Query, State},
    http::StatusCode,
};

use crate::{
    error::AppError,
    extract::{CurrentUser, MaybeUser},
    models::post::{
        CreatePostRequest, PostDetail, PostListQuery, PostPage, StarResponse, UpdatePostRequest,
    },
    services,
    state::AppState,
};

/// GET /api/posts?tag=&user=&q=&page=&limit=
pub async fn list(
    State(state): State<AppState>,
    MaybeUser(viewer): MaybeUser,
    Query(query): Query<PostListQuery>,
) -> Result<Json<PostPage>, AppError> {
    let filter = query.into_filter(None);
    let viewer_id = viewer.map(|u| u.id);
    Ok(Json(
        services::posts::list_posts(&state.pool, &filter, viewer_id).await?,
    ))
}

/// POST /api/posts — 작성 후 Discord 알림
pub async fn create(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Json(req): Json<CreatePostRequest>,
) -> Result<(StatusCode, Json<PostDetail>), AppError> {
    let post = services::posts::create_post(&state.pool, user.id, req).await?;
    services::discord::notify_new_post(&state, &post);
    Ok((StatusCode::CREATED, Json(post)))
}

/// GET /api/posts/{id}
pub async fn get(
    State(state): State<AppState>,
    MaybeUser(viewer): MaybeUser,
    Path(post_id): Path<i64>,
) -> Result<Json<PostDetail>, AppError> {
    let viewer_id = viewer.map(|u| u.id);
    Ok(Json(
        services::posts::get_post(&state.pool, post_id, viewer_id).await?,
    ))
}

/// PATCH /api/posts/{id} — 작성자만
pub async fn update(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Path(post_id): Path<i64>,
    Json(req): Json<UpdatePostRequest>,
) -> Result<Json<PostDetail>, AppError> {
    Ok(Json(
        services::posts::update_post(&state.pool, post_id, user.id, req).await?,
    ))
}

/// DELETE /api/posts/{id} — 작성자 또는 관리자
pub async fn delete(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Path(post_id): Path<i64>,
) -> Result<StatusCode, AppError> {
    services::posts::delete_post(&state.pool, post_id, &user, &state.config.upload_dir).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// PUT /api/posts/{id}/star
pub async fn star(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Path(post_id): Path<i64>,
) -> Result<Json<StarResponse>, AppError> {
    Ok(Json(
        services::posts::star(&state.pool, post_id, user.id).await?,
    ))
}

/// DELETE /api/posts/{id}/star
pub async fn unstar(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Path(post_id): Path<i64>,
) -> Result<Json<StarResponse>, AppError> {
    Ok(Json(
        services::posts::unstar(&state.pool, post_id, user.id).await?,
    ))
}
